pkg/zfs/volume.go

```diff
@@ -80,6 +83,79 @@ func init() {
 	GoogleAnalyticsEnabled = os.Getenv(GoogleAnalyticsKey)
 }
 
+func checkVolCreation(volname string) error {
+	for true {
+		vol, err := GetZFSVolume(volname)
+		if err != nil {
+			return fmt.Errorf(" wait failed, not able to get the volume %s %s", volname, err.Error())
+		}
+
+		switch vol.Status.State {
+		case ZFSStatusReady:
+			return nil
+		case ZFSStatusFailed:
+			// delete the volume
+			if err := DeleteVolume(volname); err != nil {
+				klog.Errorf(" not able to clean up the volume %s", volname)
+			}
+			return fmt.Errorf("not able to get create the volume")
+		}
+		time.Sleep(time.Second)
+	}
+	return fmt.Errorf("timeout, volume is still being created")
+}
+
+func waitForVolDestroy(volname string) error {
+	for true {
+		_, err := GetZFSVolume(volname)
+		if err != nil {
+			if k8serror.IsNotFound(err) {
+				return nil
+			}
+			return fmt.Errorf("destroy wait failed, not able to get the volume %s %s", volname, err.Error())
+		}
+		time.Sleep(time.Second)
+	}
+	return fmt.Errorf("timeout, volume is being destroyed")
+}
+
+func waitForSnapDestroy(snapname string) error {
+	for true {
+		_, err := GetZFSSnapshot(snapname)
+		if err != nil {
+			if k8serror.IsNotFound(err) {
+				return nil
+			}
+			return fmt.Errorf("destroy wait failed, not able to get the snapshot %s %s", snapname, err.Error())
+		}
+		time.Sleep(time.Second)
+	}
+	return fmt.Errorf("timeout, snapshot is being destroyed")
+}
+
+func waitForReadySnapshot(snapname string) error {
+	for true {
+		snap, err := GetZFSSnapshot(snapname)
+		if err != nil {
+			return fmt.Errorf("wait failed, not able to get the snapshot %s %s", snapname, err.Error())
+		}
+
+		switch snap.Status.State {
+		case ZFSStatusReady:
+			return nil
+		case ZFSStatusFailed:
+			// delete the snapshot
+			if err := DeleteSnapshot(snapname); err != nil {
+				klog.Errorf(" not able to clean up the snapshot %s", snapname)
+			}
+
+			return fmt.Errorf("not able to create the snapshot")
+		}
+		time.Sleep(time.Second)
+	}
+	return fmt.Errorf("timeout, snapshot is being created")
```

Comment:
q_: will this get executed? how does the timeout work?

Reply:
still experimenting it,I have two design consideration
1. block forever till the volume is created
2. have a timeout and fail the volume creation if it is not created within the timeout period.

Have not finalized it yet as there are complications involved. It is still work in progress.